### PR TITLE
[SPARK-38946][PYTHON][PS] Generates a new dataframe instead of operating inplace in setitem

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?

Generates a new dataframe instead of operating inplace in setitem

### Why are the changes needed?
Make CI passed in with pandas 1.4.3

Since pandas 1.4.0 https://github.com/pandas-dev/pandas/commit/03dd698bc1e84c35aba8b51bdd45c472860b9ec3 , dataframe.setitem should always make a copy and never write into the existing array.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
CI test with current pandas (1.3.x) and latest pandas 1.4.2, 1.4.3

### CODE REVIEW

**Comment by Yikun** [comment]
This is still a WIP PR now, only make ci passed with panda 1.4.x.

- If we want to follow pandas behavior, we might want to revert https://github.com/databricks/koalas/pull/1592
- If we only want to still keep this, we might only need this test fix PR.

**Comment by Yikun** [comment]
I haven't got enough info to see why we added https://github.com/databricks/koalas/pull/1592 before (performance or follow pandas old behavior).

cc @HyukjinKwon @ueshin Maybe you could give some input or suggestion in here.

**Comment by HyukjinKwon** [comment]
I think this needs @ueshin's look

**Comment by ueshin** [comment]
The examples in the description of https://github.com/databricks/koalas/pull/1592 are still valid.
E.g.,:

```py
>>> pd.__version__
'1.4.1'
>>> pdf = pd.DataFrame({"x": [np.nan, 2, 3, 4, np.nan, 6], "y": [np.nan, 2, 3, 4, np.nan, 6]})
>>> pser = pdf.x
>>> pser.fillna(0, inplace=True)
>>> pser
0    0.0
1    2.0
2    3.0
3    4.0
4    0.0
5    6.0
Name: x, dtype: float64
>>> pdf
     x    y
0  0.0  NaN
1  2.0  2.0
2  3.0  3.0
3  4.0  4.0
4  0.0  NaN
5  6.0  6.0
```

If we revert the change, it won't work anymore.

Have you tried to revert the changes except for the tests and run the whole tests?

If there are behavior changes in pandas, we should fix our behavior.

> I haven't got enough info to see why we added https://github.com/databricks/koalas/pull/1592 before (performance or follow pandas old behavior).

That's to follow the pandas behavior at that time and seems like it's still necessary.

**Comment by itholic** [comment]
Just confirming, any update on this ?

**Comment by Yikun** [comment]
@itholic We might want to do some special process when calling `_update_internal_frame` for pandas 1.4.x. Will update today.

**Comment by Yikun** [comment]
Test with Panda 1.4.2: https://github.com/Yikun/spark/runs/6574547715?check_suite_focus=true

After this PR, we can upgrade pandas to 1.4.2.

File: python/pyspark/pandas/frame.py
Comment (itholic) [inline python/pyspark/pandas/frame.py:504]: What if `requires_same_anchor` and `force_disconnect` are both `True` ?

**Review by itholic — APPROVED** [review]

File: python/pyspark/pandas/frame.py
Comment (itholic) [reply python/pyspark/pandas/frame.py:504]: Oh, sorry. approve is a mistake.

But looks pretty good otherwise.

Comment (Yikun) [reply python/pyspark/pandas/frame.py:504]: Thanks for review, not sure get your idea completely.

It's A or B or C, Create new dataframe if one of three coditions statified, so `if requires_same_anchor and force_disconnect are both True` also create the new dataframe.

Comment (itholic) [reply python/pyspark/pandas/frame.py:504]: Ah,,, I got it.

When I first see the name `requires_same_anchor`, I thought it means they have must same anchor between old and new `InternalFrame`, but seems like it's just for checking if they have same anchor ??

If so, how about rename the parameter to `check_same_anchor` or `verify_same_anchor` or something ?

I'm not very strong feeling about this, but I think maybe it's more proper naming.

WDYT?

Comment (Yikun) [reply python/pyspark/pandas/frame.py:504]: Yep, good suggestion, but `required_same_anchor` is orignal parameter, and already being used in current code.

Consider it's inertnal function, we could change it. I can submit a separate PR to address it to make PR easy to review.

File: python/pyspark/pandas/frame.py
Comment (xinrong-databricks) [inline python/pyspark/pandas/frame.py:5512]: Spark 3.4 is claimed to match pandas 1.4, do we need this version check?

Comment (xinrong-databricks) [reply python/pyspark/pandas/frame.py:5512]: It seems fine to keep it. I am just wondering about our principle for similar cases.

Comment (Yikun) [reply python/pyspark/pandas/frame.py:5512]: Yep, this is a good question. Now it's more for make CI passed with current pandas version 1.3.x.

I think we should remove these code after we upgrade to 1.4.x. WDYT?

File: python/pyspark/pandas/frame.py
Comment (xinrong-databricks) [inline python/pyspark/pandas/frame.py:517]: What do you think we call out `force_disconnect` takes effects on anchor frames only, or rename it with `anchor` included? My concern is that the word `Series` here may seem fuzzy.

File: python/pyspark/pandas/frame.py
Comment (xinrong-databricks) [reply python/pyspark/pandas/frame.py:5512]: Is keeping version checks in unit tests, but no version checks in main codes a good idea?

Comment (xinrong-databricks) [reply python/pyspark/pandas/frame.py:5512]: Ohh sorry didn't refresh the page to see your comment. Sounds good!

File: python/pyspark/pandas/frame.py
Comment (itholic) [reply python/pyspark/pandas/frame.py:517]: > What do you think we call out `force_disconnect` takes effects on anchor frames only, or rename it with `anchor` included? My concern is that the word `Series` here may seem fuzzy.

+1

Comment (Yikun) [inline python/pyspark/pandas/frame.py:517]:
```suggestion
        anchor_force_disconnect : bool
            Force to disconnect the original anchor and create new one
```

How about? : )

Comment (Yikun) [reply python/pyspark/pandas/frame.py:517]:
```suggestion
        anchor_force_disconnect : bool
            Force to disconnect the original anchor and create new one
```

How about? : )

File: python/pyspark/pandas/frame.py
Comment (Yikun) [reply python/pyspark/pandas/frame.py]: I think `keeping version checks in unit tests` is better, and also a better principle for similar cases:
- Don't use version check in code/implementation as far as possible, UT is a better place to do it, according UT we can easy to see what the behavior change.
- All test should be passed with any version of pandas > 1.0.5.

File: python/pyspark/pandas/frame.py
Comment (Yikun) [reply python/pyspark/pandas/frame.py]: updated. Pls let me know if you have better idea.

File: python/pyspark/pandas/frame.py
Comment (Yikun) [reply python/pyspark/pandas/frame.py]: https://issues.apache.org/jira/browse/SPARK-39310 create a jira

**Comment by xinrong-databricks** [comment]
The renaming is so much better, thanks Yikun! LGTM.

**Review by xinrong-meng — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Let me defer to @ueshin

File: python/pyspark/pandas/frame.py
Comment (ueshin) [inline python/pyspark/pandas/frame.py:8436]: We don't need to specify `requires_same_anchor` any more if `anchor_force_disconnect=True`?

File: python/pyspark/pandas/frame.py
Comment (ueshin) [inline python/pyspark/pandas/frame.py:11972]: ditto.

**Review by ueshin — COMMENTED** [review]
The condition to generate a new dataframe seems a bit more complex?

I can still see the behavior difference:

```py
>>> pd.__version__
'1.4.2'
>>> pdf = pd.DataFrame({"x": [np.nan, 2, 3, 4, np.nan, 6], "y": [np.nan, 2, 3, 4, np.nan, 6]})
>>> pser = pdf.x
>>> pdf.fillna(0, inplace=True)
>>> pdf
     x    y
0  0.0  0.0
1  2.0  2.0
2  3.0  3.0
3  4.0  4.0
4  0.0  0.0
5  6.0  6.0
>>> pser
0    0.0
1    2.0
2    3.0
3    4.0
4    0.0
5    6.0
Name: x, dtype: float64
```

whereas:

```py
>>> psdf = ps.DataFrame({"x": [np.nan, 2, 3, 4, np.nan, 6], "y": [np.nan, 2, 3, 4, np.nan, 6]})
>>> psser = psdf.x
>>> psdf.fillna(0, inplace=True)
>>> psdf
     x    y
0  0.0  0.0
1  2.0  2.0
2  3.0  3.0
3  4.0  4.0
4  0.0  0.0
5  6.0  6.0
>>> psser
0    NaN
1    2.0
2    3.0
3    4.0
4    NaN
5    6.0
Name: x, dtype: float64
```

In this case, pandas seems to not make a copy and reuse the underlying array.
I'm not sure whether this is a pandas bug or not, though.

File: python/pyspark/pandas/tests/test_dataframe.py
Comment (ueshin) [inline python/pyspark/pandas/tests/test_dataframe.py:276]: Seems like this is a common test with the old pandas?
Shall we move it out of `if LooseVersion(pd.__version__) >= LooseVersion("1.4"):`?

File: python/pyspark/pandas/tests/test_dataframe.py
Comment (ueshin) [inline python/pyspark/pandas/tests/test_dataframe.py:1460]: ditto.

File: python/pyspark/pandas/tests/test_dataframe.py
Comment (ueshin) [inline python/pyspark/pandas/tests/test_dataframe.py:2926]: ditto.

File: python/pyspark/pandas/tests/test_dataframe.py
Comment (ueshin) [inline python/pyspark/pandas/tests/test_dataframe.py:5171]: ditto.

**Comment by Yikun** [comment]
> The condition to generate a new dataframe seems a bit more complex?

@ueshin  Yes, it is really yes from your example, the `setitem` has behavior influence on several functions and change the final or part of behaviors for these functions (but pandas not mention this behavior change). Anyway, I will raise a issue on Pandas comunity to get the detail attitude for these beahavior changes.

**Comment by Yikun** [comment]
https://github.com/pandas-dev/pandas/issues/47188

**Comment by Yikun** [comment]
According to https://github.com/pandas-dev/pandas/issues/47188 and https://github.com/pandas-dev/pandas/issues/47449 , we should only address `set_item` in here, so just skip test for oldest 1.4.x pandas.

**Comment by Yikun** [comment]
@ueshin ready for review, it would be good if you could find some time. : )

**Comment by Yikun** [comment]
Let me do a brief conclusion to help review:
- **Change `setitem` to make a copy**https://github.com/pandas-dev/pandas/commit/03dd698bc1e84c35aba8b51bdd45c472860b9ec3 dataframe.setitem make a copy and never write into the existing array, so we also follow this.
- **Skip update/fillna for 1.4.0-1.4.2**: https://github.com/pandas-dev/pandas/issues/47188 this issue fix the `update` and `fillna` behaviors, I confirm it had been resolved by pandas 1.4.3, so we skip the validate copy check.
- **Skip eval for 1.4.0-1.4.3**: https://github.com/pandas-dev/pandas/issues/47449 `eval` still have a regression, so we skip it in 1.4.0~1.4.3.

I also test 1.4.2 CI with this PR: https://github.com/Yikun/spark/pull/113, all tests passed, so we can upgrade pandas to `1.4.2` in CI after this PR merged.

**Comment by HyukjinKwon** [comment]
I will leave it to @ueshin

**Comment by Yikun** [comment]
just a rebase, because we update numpy version in infra

File: python/pyspark/pandas/frame.py
Comment (Yikun) [reply python/pyspark/pandas/frame.py]: See https://github.com/apache/spark/pull/36353#issuecomment-1178956945 , after confirm from panda community, we only add `setitem` make copy. so this line had been removed

File: python/pyspark/pandas/frame.py
Comment (Yikun) [reply python/pyspark/pandas/frame.py]: See https://github.com/apache/spark/pull/36353#issuecomment-1178956945 , after confirm from panda community, we only add `setitem` make copy. so this line had been removed

File: python/pyspark/pandas/tests/test_dataframe.py
Comment (Yikun) [reply python/pyspark/pandas/tests/test_dataframe.py:1470]: See https://github.com/apache/spark/pull/36353#issuecomment-1178956945 , change the version to 1.4.0-1.4.2

File: python/pyspark/pandas/tests/test_dataframe.py
Comment (Yikun) [reply python/pyspark/pandas/tests/test_dataframe.py:2950]: See https://github.com/apache/spark/pull/36353#issuecomment-1178956945 , change the version to 1.4.0-1.4.2

File: python/pyspark/pandas/tests/test_dataframe.py
Comment (Yikun) [inline python/pyspark/pandas/tests/test_dataframe.py:1509]: Add a testcase which @ueshin mentioned before

**Comment by Yikun** [comment]
All test passed and

Test passed with Pandas 1.4.2:
https://github.com/Yikun/spark/pull/148

Test passed with Pandas 1.4.3:
https://github.com/Yikun/spark/pull/147

@ueshin Would you mind take a look? Thanks!

**Review by ueshin — APPROVED** [review]
LGTM. Thanks for working with pandas community.

**Comment by Yikun** [comment]
@ueshin Thanks for review!

**Comment by Yikun** [comment]
@HyukjinKwon Mind to take another look?

**Comment by Yikun** [comment]
Just a rebase

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
